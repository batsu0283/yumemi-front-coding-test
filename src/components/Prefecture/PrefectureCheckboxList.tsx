import type { PrefCode } from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";
import { PrefectureCheckbox } from "./PrefectureCheckbox";

type PrefectureCheckboxListProps = {
  prefectures: Prefecture[];
  checkedPrefectures: Set<PrefCode>;
  handlePrefectureCheckBox: (prefCode: PrefCode) => void;
};

export const PrefectureCheckboxList = ({
  prefectures,
  checkedPrefectures,
  handlePrefectureCheckBox,
}: PrefectureCheckboxListProps) => (
  <section>
    <h2 className="text-xl font-bold mb-4">都道府県人口</h2>
    <fieldset className="grid gap-2 grid-cols-3 sm:grid-cols-4 md:grid-cols-5">
      <legend className="sr-only">都道府県の選択</legend>
      {prefectures.map((prefecture) => (
        <PrefectureCheckbox
          key={prefecture.prefCode}
          prefCode={prefecture.prefCode}
          prefName={prefecture.prefName}
          checked={checkedPrefectures.has(prefecture.prefCode)}
          handlePrefectureCheckBox={handlePrefectureCheckBox}
        />
      ))}
    </fieldset>
  </section>
);
