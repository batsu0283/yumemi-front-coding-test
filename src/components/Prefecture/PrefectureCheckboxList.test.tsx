import { fireEvent, render, screen } from "@testing-library/react";
import type { PrefCode } from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";
import { PrefectureCheckboxList } from "./PrefectureCheckboxList";

describe("PrefectureCheckboxList", () => {
  const mockHandlePrefectureCheckBoxChange = vi.fn();

  const prefectures: Prefecture[] = [
    { prefCode: 1, prefName: "北海道" },
    { prefCode: 2, prefName: "青森県" },
    { prefCode: 3, prefName: "岩手県" },
  ];

  const checkedPrefectures = new Set<PrefCode>([1]);

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("都道府県のチェックボックスがすべて表示されること", () => {
    render(
      <PrefectureCheckboxList
        prefectures={prefectures}
        checkedPrefectures={checkedPrefectures}
        handlePrefectureCheckBoxChange={mockHandlePrefectureCheckBoxChange}
      />,
    );

    for (const prefecture of prefectures) {
      expect(
        screen.getByLabelText(`都道府県 ${prefecture.prefName} を選択`),
      ).toBeInTheDocument();
    }
  });

  it("checkedPrefecturesに基づいてチェックボックスが正しくチェックされること", () => {
    render(
      <PrefectureCheckboxList
        prefectures={prefectures}
        checkedPrefectures={checkedPrefectures}
        handlePrefectureCheckBoxChange={mockHandlePrefectureCheckBoxChange}
      />,
    );

    expect(screen.getByLabelText("都道府県 北海道 を選択")).toBeChecked();
    expect(screen.getByLabelText("都道府県 青森県 を選択")).not.toBeChecked();
    expect(screen.getByLabelText("都道府県 岩手県 を選択")).not.toBeChecked();
  });

  it("チェックボックスがクリックされた時に handlePrefectureCheckBoxChange が呼ばれること", () => {
    render(
      <PrefectureCheckboxList
        prefectures={prefectures}
        checkedPrefectures={checkedPrefectures}
        handlePrefectureCheckBoxChange={mockHandlePrefectureCheckBoxChange}
      />,
    );

    const checkbox = screen.getByLabelText("都道府県 岩手県 を選択");

    fireEvent.click(checkbox);
    expect(mockHandlePrefectureCheckBoxChange).toHaveBeenCalledWith(3);
    expect(mockHandlePrefectureCheckBoxChange).toHaveBeenCalledTimes(1);
  });

  it("Snapshot", () => {
    const { container } = render(
      <PrefectureCheckboxList
        prefectures={prefectures}
        checkedPrefectures={checkedPrefectures}
        handlePrefectureCheckBoxChange={mockHandlePrefectureCheckBoxChange}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
