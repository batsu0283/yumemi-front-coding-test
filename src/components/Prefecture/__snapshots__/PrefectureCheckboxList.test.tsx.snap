// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefectureCheckboxList > Snapshot 1`] = `
<div>
  <section>
    <h2
      class="text-xl font-bold mb-4"
    >
      都道府県人口
    </h2>
    <fieldset
      class="grid gap-2 grid-cols-3 sm:grid-cols-4 md:grid-cols-5"
    >
      <legend
        class="sr-only"
      >
        都道府県の選択
      </legend>
      <label
        class="flex items-center gap-2 p-2 bg-gray-100 rounded-lg shadow-sm"
      >
        <input
          aria-label="都道府県 北海道 を選択"
          checked=""
          class="w-5 h-5 text-cyan-600 focus:ring-cyan-500"
          type="checkbox"
        />
        <span
          class="text-gray-700"
        >
          北海道
        </span>
      </label>
      <label
        class="flex items-center gap-2 p-2 bg-gray-100 rounded-lg shadow-sm"
      >
        <input
          aria-label="都道府県 青森県 を選択"
          class="w-5 h-5 text-cyan-600 focus:ring-cyan-500"
          type="checkbox"
        />
        <span
          class="text-gray-700"
        >
          青森県
        </span>
      </label>
      <label
        class="flex items-center gap-2 p-2 bg-gray-100 rounded-lg shadow-sm"
      >
        <input
          aria-label="都道府県 岩手県 を選択"
          class="w-5 h-5 text-cyan-600 focus:ring-cyan-500"
          type="checkbox"
        />
        <span
          class="text-gray-700"
        >
          岩手県
        </span>
      </label>
    </fieldset>
  </section>
</div>
`;
